File: CSES/Range Queries/ForestQueriesII.java
import java.util.*;
import java.io.*;
 
public class ForestQueriesII {
 
	// https://cses.fi/problemset/task/1739/
	
	public static void main(String[] args) throws IOException, FileNotFoundException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		//BufferedReader in = new BufferedReader(new FileReader("ForestQueriesII"));
 
		StringTokenizer st = new StringTokenizer(in.readLine());
		int n = Integer.parseInt(st.nextToken());
		int q = Integer.parseInt(st.nextToken());
		char[][] arr = new char[n][];
		BIT2D bit = new BIT2D(n,n);
		for (int i=0; i<n; i++) {
			arr[i] = in.readLine().toCharArray();
			for (int j=0; j<n; j++) {
				if (arr[i][j] == '*') bit.set(i,j,1);
			}
		}
		
		StringBuilder s = new StringBuilder();
		for (int i=0; i<q; i++) {
			st = new StringTokenizer(in.readLine());
			int c=Integer.parseInt(st.nextToken());
			if (c == 2) {
				int one=Integer.parseInt(st.nextToken())-1;
				int two=Integer.parseInt(st.nextToken())-1;
				int three=Integer.parseInt(st.nextToken())-1;
				int four=Integer.parseInt(st.nextToken())-1;
				s.append(bit.sum(one,two,three,four));
				s.append("\n");
			}
			else {
				int one=Integer.parseInt(st.nextToken())-1;
				int two=Integer.parseInt(st.nextToken())-1;
				if (arr[one][two] == '.') {
					arr[one][two] = '*';
					bit.set(one, two, 1);
				}
				else {
					arr[one][two] = '.';
					bit.set(one, two, -1);
				}
			}
		}
		System.out.print(s);
	}
	
	static class BIT2D {
		int n,m;
		long[][] f;		// 1 base indexing
		BIT2D(int n, int m) {
			this.n = n; this.m = m; f = new long[n+1][m+1];
		}
		
		// sum from (i,j) to (0,0)
		long sum (int i, int j) {
			i++; j++;
			long ret=0;
			while (i>0) {
				int y=j;
				while (y>0) {
					ret += f[i][y];
					y -= y&-y;
				}
				i -= i&-i;
			}
			return ret;
		}
		
		// sum from (i1, j1) to (i2, j2)
		long sum(int i1, int j1, int i2, int j2) {
			return sum(i2, j2) - sum(i1-1, j2) - sum(i2, j1-1) + sum(i1-1, j1-1);
		}
		
		// add value to (i,j)
		void set(int i, int j, long val) {
			i++; j++;
			while (i<=n) {
				int y=j;
				while (y<=m) {
					f[i][y] += val;
					y += y&-y;
				}
				i += i&-i;
			}
		}
		
		// add value from (i1, j1) to (i2, j2) --> arr[i][j] = sum(i, j);
		void range_set(int i1, int j1, int i2, int j2, long val) {
			set(i1, j1, val);
			set(i1, j2+1, -val);
			set(i2+1, j1, -val);
			set(i2+1, j2+1, val);
		}
	}
}

File: CSES/Range Queries/ForestQueriesII.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;

// https://cses.fi/problemset/task/1739

const int MaxN = 1e3+10;
const int MaxM = 1e3+10;

struct BIT2D {
    int n,m;
    ll f[MaxN][MaxM];		// 1 base indexing
    
    BIT2D (int x, int y) {
        n = x; m = y;
        for (int i=0; i<=n; i++) {
            for (int j=0; j<=m; j++) {
                f[i][j] = 0;
            }
        }
    }

    // sum from (i,j) to (0,0)
    ll sum (int i, int j) {
        i++; j++;
        ll ret=0;
        while (i>0) {
            int y=j;
            while (y>0) {
                ret += f[i][y];
                y -= y&-y;
            }
            i -= i&-i;
        }
        return ret;
    }
    
    // sum from (i1, j1) to (i2, j2)
    ll sum(int i1, int j1, int i2, int j2) {
        return sum(i2, j2) - sum(i1-1, j2) - sum(i2, j1-1) + sum(i1-1, j1-1);
    }
    
    // add value to (i,j)
    void add(int i, int j, ll val) {
        i++; j++;
        while (i<=n) {
            int y=j;
            while (y<=m) {
                f[i][y] += val;
                y += y&-y;
            }
            i += i&-i;
        }
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, q;
    cin >> n >> q;
    BIT2D bit(n, n);
    char arr[n][n];
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            cin >> arr[i][j];
            if (arr[i][j] == '*') bit.add(i, j, 1);
        }
    }

    for (int i=0; i<q; i++) {
        int t; cin >> t;
        if (t == 1) {
            int x, y;
            cin >> x >> y;
            x--; y--;
            if (arr[x][y] == '*') {
                bit.add(x, y, -1);
                arr[x][y] = '.';
            }
            else {
                bit.add(x, y, 1);
                arr[x][y] = '*';
            }
        }
        else {
            int x1, y1, x2, y2;
            cin >> x1 >> y1 >> x2 >> y2;
            x1--; y1--; x2--; y2--;
            cout << bit.sum(x1, y1, x2, y2) << "\n";
        }
    }
}